javascript/ql/src/semmle/javascript/security/dataflow/IncompleteHtmlAttributeSanitizationCustomizations.qll

```diff
@@ -0,0 +1,90 @@
+/**
+ * Provides default sources, sinks and sanitizers for reasoning about
+ * incomplete HTML sanitization vulnerabilities, as well as extension
+ * points for adding your own.
+ */
+
+import javascript
+import semmle.javascript.security.dataflow.RemoteFlowSources
+import Security.IncompleteBlacklistSanitizer
+
+module IncompleteHtmlAttributeSanitization {
+  /**
+   * A data flow source for incomplete HTML sanitization vulnerabilities.
+   */
+  abstract class Source extends DataFlow::Node {
+    /**
+     * Gets a character that may come out of this source.
+     */
+    abstract string getAnUnsanitizedCharacter();
+  }
+
+  /**
+   * A data flow sink for incomplete HTML sanitization vulnerabilities.
+   */
+  abstract class Sink extends DataFlow::Node {
+    /**
+     * Gets a character that is dangerous for this sink.
+     */
+    abstract string getADangerousCharacter();
+  }
+
+  /**
+   * A sanitizer for incomplete HTML sanitization vulnerabilities.
+   */
+  abstract class Sanitizer extends DataFlow::Node { }
+
+  /**
+   * A source of incompletely sanitized characters, considered as a
+   * flow source for incomplete HTML sanitization vulnerabilities.
+   */
+  class IncompleteHtmlSanitizerAsSource extends Source, HtmlSanitization::IncompleteSanitizer {
+    override string getAnUnsanitizedCharacter() {
+      result = HtmlSanitization::IncompleteSanitizer.super.getAnUnsanitizedCharacter()
+    }
+  }
+
+  /**
+   * A concatenation that syntactically looks like a definition of an HTML attribute.
+   */
+  private class HtmlAttributeConcatenation extends DataFlow::Node {
+    string lhs;
+
+    HtmlAttributeConcatenation() {
+      exists(StringOps::ConcatenationLeaf n |
```

Comment:
Two suggestions:
- We can safely subclass `StringOps::ConcatenationLeaf` instead of `n = this` (it's not abstract)
- Possibly even subclass `StringOps::HtmlConcatenationLeaf` which additionally checks that this is part of an HTML tag. I'm not sure it's needed, but worth thinking about. The escaping of `<` is a strong indicator that `&` must be escaped too, and the preceding `"` is a strong indicator that quotes must be escaped, so feel free to ignore if you haven't seen any spurious sinks.

Reply:
The correlation between the HTML-sanitizer and the sink are strong enough that we avoid false positives. I even think `StringOps::HtmlConcatenationLeaf` is a bit too restrictive since it would cost us this real-world example: ` ' ' + a.name + '="' + escapeHTML(a.value).replace(/"/g, '&quot;') + '"';`